File: aoc/python/thirteen.py
from __future__ import annotations
from typing import TYPE_CHECKING
from dataclasses import dataclass
from copy import deepcopy
import itertools
if TYPE_CHECKING:
    GRID = list[list[bool]]

TEST_INPUT = """6,10
0,14
9,10
0,3
10,4
4,11
6,0
6,12
4,1
0,13
10,12
3,4
3,0
8,4
1,10
2,14
8,10
9,0

fold along y=7
fold along x=5
"""


@dataclass
class Fold:
    position: int
    along_x: bool


def parse_input(input: str) -> tuple[GRID, list[Fold]]:
    p_str, ins_str = input.split("\n\n")

    instructions = [Fold(int(num), loc == 'x') for (loc, num) in map(
        lambda s: s.split(" ")[2].split("="), ins_str.splitlines())]

    points = [(int(x), int(y)) for x, y in map(lambda s: s.split(","), p_str.splitlines())]

    maxX = max(x for x, _ in points)
    maxY = max(y for _, y in points)

    grid = []

    for i in range(maxY+1):
        grid.append([False]*(maxX+1))

    for col, row in points:
        grid[row][col] = 1

    return (grid, instructions)


NUMS = {0: '.', 1: '#'}


def print_grid(g: GRID):
    for row in g:
        print(''.join(NUMS[x] if x in NUMS else str(x) for x in row))


def fold_grid(g: GRID, instruction: Fold) -> GRID:
    to_fold = g
    if instruction.along_x:
        left, right = (
            [r[:instruction.position] for r in to_fold],
            [r[instruction.position+1:] for r in to_fold]
        )

        for r in right:
            r.reverse()

        for row in range(len(right)):
            for col in range(len(right[row])):
                left[row][col] |= right[row][col]

        to_fold = left

    else:
        top, bottom = to_fold[:instruction.position], to_fold[instruction.position+1:]
        bottom.reverse()

        for row in range(len(bottom)):
            for col in range(len(bottom[row])):
                top[row][col] |= bottom[row][col]

        to_fold = top

    return to_fold


def part_1(input: str) -> str:
    g, i = parse_input(input)
    g = fold_grid(g, i[0])

    return f'{sum(itertools.chain.from_iterable(g))}'


def part_2(input: str) -> str:
    g, i = parse_input(input)

    for instruction in i:
        g = fold_grid(g, instruction)

    print_grid(g)

    return 'see above'

File: aoc/rust/src/days/thirteen.rs
use itertools::Itertools;

#[derive(Debug)]
struct Instruction {
    num: usize,
    along_x: bool,
}

type GRID = Vec<Vec<bool>>;

fn parse_input(input: &String) -> (GRID, Vec<Instruction>) {
    let (top, bottom) = input.split("\n\n").collect_tuple().unwrap();
    let points: Vec<(usize, usize)> = top
        .lines()
        .map(|l| {
            let (x, y) = l.split(",").collect_tuple().unwrap();
            (x.parse().unwrap(), y.parse().unwrap())
        })
        .collect_vec();

    let max_x = points.iter().map(|t| t.0).max().unwrap();
    let max_y = points.iter().map(|t| t.1).max().unwrap();
    let mut grid = Vec::new();
    for _ in 0..max_y + 1 {
        let mut new_row = Vec::new();
        new_row.resize(max_x + 1, false);

        grid.push(new_row);
    }

    for (col, row) in points {
        grid[row][col] = true;
    }

    // and now for instructions

    let instructions = bottom
        .lines()
        .map(|l| {
            l.split(" ").collect_vec()[2]
                .split("=")
                .collect_tuple()
                .unwrap()
        })
        .map(|(d, n)| Instruction {
            num: n.parse().unwrap(),
            along_x: d == "x",
        })
        .collect_vec();

    (grid, instructions)
}

fn print_grid(g: &GRID) {
    let mut lines = String::new();
    for r in g {
        lines.push_str(&r.iter().map(|&b| if b { '#' } else { '.' }).join(""));
        lines.push('\n');
    }

    lines.push('\n');
    println!("{}", lines)
}

fn fold(g: GRID, instruction: &Instruction) -> GRID {
    if instruction.along_x {
        let left = g
            .iter()
            .map(|r| (&r[(instruction.num + 1)..]).to_vec())
            .collect_vec();
        let right = g
            .iter()
            .map(|r| {
                let mut v = (&r[..instruction.num]).to_vec();
                v.reverse();
                v
            })
            .collect_vec();

        right
            .iter()
            .zip(left)
            .map(|(x, y)| x.iter().zip(y).map(|(&x, y)| x || y).collect_vec())
            .collect_vec()
    } else {
        let top = g[..instruction.num].to_vec();
        let mut bottom = g[(instruction.num + 1)..].to_vec();

        bottom.reverse();
        top.iter()
            .zip(&bottom)
            .map(|(x, y)| x.iter().zip(y).map(|(&a, &b)| a || b).collect_vec())
            .collect_vec()
    }
}

pub fn part_1(input: &String) -> usize {
    // let input = &TEST_INPUT.to_owned();
    let (g, instructions) = parse_input(input);
    let g = fold(g, &instructions[0]);
    g.iter().flatten().filter(|&&x| x).count()
}

pub fn part_2(input: &String) -> usize {
    let (mut g, instructions) = parse_input(input);

    for i in &instructions {
        g = fold(g.clone(), i);
    }

    // No I dont know why its reversed
    g = g
        .iter()
        .map(|x| ((*x).clone()).into_iter().rev().collect_vec())
        .collect_vec();

    print_grid(&g);

    0
}
